include/mbgl/perf/runtime_metrics.hpp

```diff
@@ -0,0 +1,201 @@
+
+struct TraceInterval {
+    unsigned char scope = 0;    // overwritten by each range
+    uint64_t begin = 0u;        // overwritten by each range
+    uint64_t end = 0u;          // overwritten by each range
+    uint64_t totalMs = 0u;
+    uint64_t totalCalls = 0u;
+};
+
+#define STRUCT_BEGIN(name) struct name {
+#define STRUCT_END() };
+#define STRUCT_MEMBER_UINT64_RANGE(name, store_range, store_counter)   TraceInterval name;
+#include <mbgl/perf/trace_data.in>
+#undef STRUCT_BEGIN
+#undef STRUCT_END
+#undef STRUCT_MEMBER_UINT64_RANGE
+
+// Instrumentation-less tracing implementation, possibly requiring mutexing
+class Tracer
+{
+public:
+    static uint64_t nanoseconds() {
+        return Clock::now().time_since_epoch().count();
+    }
+    static Tracer& get() {
+        static Tracer i; // Guaranteed to be destroyed.
+        // Instantiated on first use.
+        return i;
+    }
+private:
+    Tracer() {}
+
+public:
+    Tracer(Tracer const&)               = delete;
+    void operator=(Tracer const&)       = delete;
+    bool collectData = false;
+
+
+    std::mutex frame_mutex;
+    std::vector<mbgl::util::FrameTrace> frameList;
+    mbgl::util::FrameTrace frameTrace;
+    unsigned int currentScope = 0u;
+    unsigned int appStart = 0u;
+
+/* frameTraceReset */
+#define STRUCT_BEGIN(name) static void frameTraceReset(name &data) { data = name();  // clear it by assigning to a new struct
+#define STRUCT_END() };
+#define STRUCT_MEMBER_UINT64_RANGE(name, store_range, store_counter)
+#include <mbgl/perf/trace_data.in>
+#undef STRUCT_BEGIN
+#undef STRUCT_END
+#undef STRUCT_MEMBER_UINT64_RANGE
+
+/* framesToJson */
+#define STRUCT_BEGIN(name) \
+static std::string frameToJson(const name &data) { \
+    std::string res; \
+    res += "{\n";
+// The following assumes that one of { store_range, store_counter } is true.
+#define STRUCT_MEMBER_UINT64_RANGE(name, store_range, store_counter) \
+res += std::string("\"") + std::string(#name) + std::string("\" : { "); \
+if (store_range) { \
+    res += std::string("\"scope\" : ") + std::to_string(data.name.scope) + std::string(" , ") \
+    + std::string("\"begin\" : ") + std::to_string(data.name.begin) + std::string(" , ") \
+    + std::string("\"end\" : ") + std::to_string(data.name.end); } \
+if (store_counter) { if (store_range) res += std::string(","); \
+    res += std::string("\"totalMs\" : ") + std::to_string(data.name.totalMs) + std::string(" , ") \
+    + std::string("\"totalCalls\" : ") + std::to_string(data.name.totalCalls); \
+} \
+res += std::string(" },");
+#define STRUCT_END() \
+    res.pop_back(); res += "}"; \
+    return res; \
+};
+#include <mbgl/perf/trace_data.in>
+#undef STRUCT_BEGIN
+#undef STRUCT_END
+#undef STRUCT_MEMBER_UINT64_RANGE
+
+    inline bool isCollectingData() const {
+        return collectData;
+    }
+    inline void setCollectData(bool enabled) {
+        collectData = enabled;
+    }
+
+    inline std::string framesJson(bool steal = false) const {
```

Comment:
with MBGL_TRACE_COLLECT(path) we could set a flag that will dump tree as soon as stack level goes to 0, e.g., outermost MBGL_TRACE_END is invoked.

Also, we should avoid using custom serializers, since we already have rapidjson as a part of a binary. We could export collected events with 4 lines of code.
```c++
rapidjson::StringBuffer s;
rapidjson::Writer<rapidjson::StringBuffer> writer(s);
stringify(writer, MapOfEvents);
mbgl::write_file("results.json", s.GetString());
```